### PR TITLE
killmessages: refactor

### PR DESCRIPTION
**Pull request by iDannz1**
As requested in #331, I was trying to fix the CPU leak problem, but in fact, after some minutes I decided to just rewrite it from scratch, in a much simpler way.

I tested it with [this](https://pastebin.com/raw/LtvP7uz2) and this was the biggest CPU consumption for me with 15 messages:
image
(_I didn't test with other players, if you wanna help to test contact me in discord_)

All parameters of exported functions were left in the same order, to do not break anything, except the function `setKillMessageStyle` that was removed because I also added the [MTA settings system](https://wiki.multitheftauto.com/wiki/Settings_system) (get) to set up the following settings: `displayLines`, `drawMarginRight`, `drawMarginBottom`, `fadeTime`, `duration`, and `outputConsole`. So now it can be changed using other resources, like `admin`.

Anyways, please review carefully before merging, because im a little afraid of breaking some existing scripts (or forgetting some other functionality), let me know if you want me to change anything.

### CODE REVIEW

**Comment by Dutchman101** [comment]
The killfeed isn't centered on right side of screen like before, maybe you should make it match?

It's quite a bit lower.. at least on my resolution 1920 x 1080

**Comment by Dutchman101** [comment]
> I think this resource deserves a rewrite, mainly because of this CPU usage and also because its code looks a little spaghetti. I can take a look in some hours...

we ought to implement vehicle kill detection for killers in hydra/hunter/rustler/rhino etc.

_Originally posted by @ricksterhd123 in https://github.com/multitheftauto/mtasa-resources/issues/331#issuecomment-868689923_

^ I agree with that.. for reference, MTA doesn't let you detect the killer/victim *inside* a vehicle when primary rockets of a vehicle like hydra/hunter are used.

But barely anything is impossible, and this also wasn't. You can work around it by 'guessing' the killer with a smart script, and in practise i've observed an accuracy rate of 99.5%.. it records all vehicle-on-vehicle kills accurately and rarely (if ever) misidentifies a killer/victim.

See example implementation of what im describing in this customized "support hydra / hunter" kills of `killmessages` resource: [https://community.mtasa.com/index.php?p=resources&s=details&id=14801](https://community.mtasa.com/index.php?p=resources&s=details&id=14801)

I suggest you use a similar technique to add support for vehicle kills to your refactored killmessages @iDannz1

**Comment by iDannz1** [comment]
> See example implementation of what im describing in this customized "support hydra / hunter" version of killmessages resource: https://community.mtasa.com/index.php?p=resources&s=details&id=14801

I added the changes suggested above using your version as inspiration, thank you.

Also, thank you @androksi for helping with testing. 
(We just forgot to test with Rhyno, so please make sure it's working before merging)

**Comment by ricksterhd123** [comment]
I will test this later today, thank you.

**Comment by ricksterhd123** [comment]
Rhino works nicely

**Comment by jlillis** [comment]
Has anyone been able to test this in an environment with many players? It all looks good but I can't confirm any performance improvements since it's just me testing locally.

**Comment by ricksterhd123** [comment]
I think dutchman101 is running this version on his server. It's been working fine so far as far as I know.

Cheers

Ricky
________________________________

**Comment by Dutchman101** [comment]
Yeah, it's been tested on a large scale without any bugs found - just has a quirk that im not sure of whether it's bad or not.

To best put it into words.. 

The old killmessages draws new kills in at the top, letting new kills build below the 'first' one - then fading out the oldest ones at the top.

The new killmessages (this PR) draws new kills in at the bottom, making new kills appear above it (towards the top) and additionally also times out ones that are listed for too long, in the absence of new kills to fade them out forcibly.

So, the way killfeed works is sort of reversed. It looks good, but im worried about what people got used to.. it will cause confusion. People are used to look at the bottom if they want to see the latest kills. What do we do? Keep feature parity or say it's fine?

**Comment by iDannz1** [comment]
> Yeah, it's been tested on a large scale without any bugs found - just has a quirk that im not sure of whether it's bad or not.
> 
> To best put it into words..
> 
> The old killmessages draws new kills in at the top, letting new kills build below the 'first' one - then fading out the oldest ones at the top.
> 
> The new killmessages (this PR) draws new kills in at the bottom, making new kills appear above it (towards the top) and additionally also times out ones that are listed for too long, in the absence of new kills to fade them out forcibly.
> 
> So, the way killfeed works is sort of reversed.. it's an entirely new flow. It looks good, but im worried about what people got used to.. it will cause confusion. People are used to look at the bottom if they want to see the latest kills. What do we do? Keep feature parity or say it's fine?

It looks pretty easy to fix, tomorrow morning I can add a new commit making it look like you described, I haven't really realized that this is the order in which the old killmessages work 😅.

File: [gameplay]/killmessages/scripts/client/c.ui.lua [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                
+                local killerWidth = dxGetTextSize(killerText, 0, 1, 1, font, false, true)
+                
+                x = x - padding - killerWidth
+                
+                dxDrawText(killerText, x + 2, y + 2, x + killerWidth, y + lineHeight, tocolor(0, 0, 0, alpha), 1, font, 'center', 'center')
+                dxDrawText(killerText, x, y, x + killerWidth, y + lineHeight, killerColor, 1, font, 'center', 'center')
+            end
+            
+            y = y - lineHeight - padding
```
Comment (iDannz1) [inline [gameplay]/killmessages/scripts/client/c.ui.lua:102]:
```suggestion
            y = y + lineHeight + padding
````

**Pull request closed without merging**